arroyo/processing/processor.py

```diff
@@ -218,24 +261,36 @@ def _run_once(self) -> None:
                             logger.info(
                                 "Paused for longer than %d seconds", LOG_THRESHOLD_TIME
                             )
+
+                        if paused_duration is not None:
+                            self.__metrics_buffer.add_pause_time(paused_duration)
+
                 else:
                     # If we were trying to submit a message that failed to be
                     # submitted on a previous run, we can resume accepting new
                     # messages.
                     if message_carried_over:
                         assert self.__paused_timestamp is not None
                         paused_duration = time.time() - self.__paused_timestamp
-                        self.__paused_timestamp = None
-                        self.__last_log_timestamp = None
                         logger.debug(
                             "Successfully submitted %r, resuming consumer after %0.4f seconds...",
                             self.__message,
                             paused_duration,
                         )
                         self.__consumer.resume([*self.__consumer.tell().keys()])
-                        self.__metrics.timing(
-                            "pause_duration_ms", paused_duration * 1000
+                        last_recorded = (
+                            self.__last_log_timestamp
+                            if self.__last_log_timestamp is not None
+                            else self.__paused_timestamp
+                        )
+
+                        self.__metrics_buffer.add_pause_time(
+                            time.time() - last_recorded
                         )
+
+                        self.__paused_timestamp = None
+                        self.__last_log_timestamp = None
```

Comment:
I am not sure you need this part. 
It only meaningful covers the resume call. Logging the interval between the current time and __last_log_timestamp or __paused_timestamp would overcounting for the same reason I mentioned above.

Reply:
It is needed. Otherwise anything under 5 seconds won't get counted

Reply:
I don't think this is correct. 
The loop works like this:
- get a message from kafka 
- submit to strategy
- MessageRejected
- pause consumer
- keep trying to submit to strategy as long as it keeps raising the exception

There is no pause or sleep in that loop.
So you will keep submitting and receiving the exceptions thousands of times per second.
So the `except` block keeps getting executed thousands of times per second. All these variables `pause_duration`, `self.__paused_timestamp`, `self.__last_log_timestamp` were only there in order to log every 5 seconds how long we have been paused.

I think you can remove all that complexity and simply accumulate pause time every time you go through the loop. So you flush the metric every second.

Reply:
This is the part of the code where the consumer get resumed. I think it's important that the last pause metric gets recorded here immediately after consumer.resume() is called.. I needed to build it this way before to make both logging and metrics work together. Since we've decided to drop the log, I think this has been simplified quite a bit now.